### ISSUE TITLE
Agent-driven simplified modeling: rux MCP gateway + headless rendering + Blender workflow

### ISSUE THREAD

**Issue by pfmephisto** [pfmephisto/ReUseX#267]
# Agent-driven simplified modeling: rux MCP gateway + headless rendering + Blender (blender-mcp) workflow

## Goal

Evaluate and prototype an **agent-driven alternative to the geometric reconstruction pipeline**: an LLM agent with full access to ReUseX project data (via a new **rux MCP gateway**) drives Blender (via **blender-mcp**) to construct a simplified building model and a tagged component inventory — as an alternative/complement to the current planes → rooms → CellComplex/MIP → windows pipeline, whose results have so far been underwhelming.

Inspiration: OpenAI's [architectural visualization with Astra](https://developers.openai.com/blog/architectural-visualization-with-astra) — but note the reality check below: that workflow is an agentic `bpy`-code loop, not native 3D generation, which means the pattern is **model-agnostic and reproducible** with any strong agentic model. What we must build is the *harness*: data access, agent "eyes", and the Blender bridge.

### What Astra actually does
- GPT-6 Astra (2026-09-04) is a text+image → text agentic model. **No native 3D input/output** — no point clouds, no meshes.
- Its architectural showcase = Astra in a Codex environment writing Blender Python (`bpy`) in an iterative render-inspect-refine loop; export to UE5. Inputs were a text brief + reference images.
- Its 95.9% "BenchCAD" score is a *CAD code generation* benchmark (⚠️ not independently audited).
- Implication: the differentiator is not the model, it's the **loop**: structured scene evidence in → code actions out → rendered feedback in. We can build that loop around Claude (or any model) today.

### State of the art: agents building 3D scenes
- All successful systems (SceneCraft, 3D-GPT, Holodeck 2.0, BIMgent) use **iterative loops with visual feedback**; one-shot generation fails for architecture.
- **Multi-view renders are the proven interface for agent spatial grounding** (SPAZER 2506.21924, S-Agent 2606.20515, SpatialAct, Think3D): render 3–6 viewpoints → vision model reasons → acts → re-render.
- **No published system does LLM-agent scan-to-simplified-BIM end-to-end.** BIMgent (2506.07217) drives Revit's GUI from floor plans at 32% task success. Scan-to-BIM in industry remains DL-segmentation + parametric fitting + human QA (A-Scan2BIM: "assistive, not autonomous").
- **Our position is unusually strong for a hybrid**: ReUseX already extracts planes (normals/centroids), rooms (visibility graph + Leiden), instances, and component candidates geometrically. The agent doesn't interpret raw scans — it works from *structured geometric evidence* plus renders, and applies judgment where the MIP/heuristics currently fail (topology decisions, component classification, simplification choices).

### blender-mcp landscape
- **ahujasid/blender-mcp** (community standard, Blender 3.0+): scene inspection, object create/modify, materials, asset libraries, and — critically — **arbitrary `bpy` Python execution** (the escape hatch that makes everything else possible). Geometry import: FBX/OBJ/GLB/STL/PLY. No point-cloud tools (fine — see architecture: Blender never holds the raw scan).
- **Official Blender Foundation ↔ Anthropic connector** (2026-04, "Claude for Creative Work", Blender 4.2+/5.1+ — ⚠️ min version inconsistent across sources, verify): full `bpy` API access via Claude Desktop.
- Notable fork: **PatrykIti/blender-ai-mcp** — vision-assisted feedback loops, deterministic verification, layout macros; most relevant to architectural use.
- Known limits: no parametric constraint solver (exact wall layouts need explicit Python), boolean ops make messy topology, per-object round-trips get slow on complex scenes.
- Tagging: custom properties (`obj["guid"] = …`) via the Python escape hatch — sufficient for a tagged inventory.

### Exposing large 3D data to agents (MCP patterns)
- Reference architecture for MCP-over-BIM (arXiv 2601.00809): **progressive disclosure** — lightweight summaries as resources, queries as tools, geometry fetched on demand, *never raw point clouds in context*.
- Survey of existing MCP servers (Speckle, Revit ×4 incl. official Autodesk, Rhino/Grasshopper, QGIS): all expose metadata/queries/operations; **none expose raw point clouds** — server-side abstraction is universal.
- **No MCP server for point-cloud data exists off the shelf** — this gateway is novel but follows established patterns.
- C++ MCP SDK status: **no official SDK** (TS + Python only); community C++ options exist but lag the spec. The proven pattern for wrapping an existing native app (pyRevit precedent): **thin Python FastMCP server shelling out to the CLI.**

## Current assets & gaps (verified in-repo)

**Assets — we're closer than expected:**
| Asset | Where | Agent relevance |
|---|---|---|
| `rux get <path>` | `apps/rux/src/get.cpp` | Path-addressable JSON/binary queries over the whole DB: `clouds`, `frames.ID.pose/intrinsics/color/depth`, `components`, `passports`, `meshes.NAME.data`, `log` — **this is already 80% of an MCP tool surface** |
| `rux info --json` | `apps/rux/src/info.cpp` | Project summary as JSON |
| `rux analyze quality/accuracy` | `apps/rux/src/analyze/` | JSON metrics — usable as **agent-loop validation** (did the simplified model improve?) |
| `rux export csv / materialepas / ply / e57 / rhino / speckle` | `apps/rux/src/export/` | Component inventory + geometry hand-off |
| pybind11 bindings | `bindings/python/` (BUILD_PYTHON_BINDINGS=OFF) | Read-only ProjectDB access incl. `reconstruct_frame(s)`; written but disabled — a Python MCP server could use these instead of subprocess for bulk data |
| ruxd | `apps/ruxd/` | 7 routes, `/segment/planes` is a stub — **not** a dependency for this work; stays headless per #265 |

**Gaps:**
1. **No headless rendering.** `rux view` requires a display; no `--screenshot`/offscreen mode exists. This is the critical missing piece — the agent's "eyes".
2. **No MCP surface** anywhere in the repo.
3. Bindings expose component *counts/names* only — not full component geometry (`building_component` singular is C++-only today).
4. No write-back path for agent results (tagged model → ProjectDB `building_components`).

### Lessons from the Blender addon test (maintainer feedback, 2026-09)
The existing `apps/blender/` addon was a test and **not successful**: loading each sensor frame as an individual object makes Blender slow, and the desired editing controls could not be built into Blender's UI. Consequences for this design:
- **Blender holds only the simplified model being constructed** — never raw scan data, never per-frame objects. The agent sees the scan via server-side renders from the MCP gateway, not by importing clouds into Blender.
- At most ONE decimated reference cloud (or a GLB of detected plane patches) is imported for visual alignment, and only if the workflow proves to need it.
- Interactive *human* editing controls are out of scope for Blender — that belongs to the GUI workstream (#265). Blender is the **agent's modeling tool**, not the user's editor.

## Proposed architecture

```
┌────────────────────────── agent (Claude / any agentic model) ──────────────────────────┐
│   iterative loop: query evidence → view renders → model in Blender → verify → refine   │
└──────────┬─────────────────────────────────────────────────────┬───────────────────────┘
           │ MCP                                                 │ MCP
┌──────────┴──────────────────────────┐            ┌─────────────┴────────────────┐
│  rux MCP gateway (NEW)              │            │  blender-mcp (existing OSS)  │
│  thin Python FastMCP server         │            │  scene ops + bpy escape      │
│  wraps `rux` CLI (+ pybind11 later) │            │  hatch; custom-property      │
│  ── resources ──                    │            │  tagging; GLB import         │
│  project summary, label defs,       │            └─────────────┬────────────────┘
│  coordinate system, inventory       │                          │
│  ── tools ──                        │                 ┌────────┴────────┐
│  query_planes/rooms/components      │                 │  Blender        │
│  get_frame(color|depth|pose)        │                 │  simplified     │
│  render_views(cloud, viewpoints[])  │◀── NEW rux      │  model + tags   │
│  export_geometry(→ GLB)             │    render       └────────┬────────┘
│  analyze_quality / compare          │    subcommand            │ export GLB/IFC +
│  save_components(tagged inventory)  │◀─────────────────────────┘ inventory JSON
└──────────┬──────────────────────────┘   write-back
           │ subprocess / bindings
     ProjectDB (.rux)
```

**Design rules:**
1. **Progressive disclosure**: summaries → queries → renders → geometry exports. Raw clouds never enter agent context.
2. **The gateway is a thin adapter** over existing `rux` capabilities (`get`, `info --json`, `analyze`, `export`, new `render`). New capability goes into `rux`/`libs/reusex` first, gateway second — consistent with the library-first rule and the #265 contract philosophy.
3. **Agent evidence package** per session: plane inventory (normal/centroid/extent/label), room graph, instance/component candidates with bboxes + confidence, floor-plan-style top-down render, elevation renders, quality metrics. This is "as much data as possible" in *useful* form.
4. **Round-trip validation**: after the agent builds the simplified model, export it and compare against the point cloud (`analyze accuracy`-style distance metrics) — gives the loop an objective error signal, not just visual judgment.
5. **Output contract**: simplified model (GLB/IFC + Blender file) + tagged inventory written back to `building_components` (guid-linked, `confidence` field marks agent-generated entries) so downstream export (Rhino/Speckle/CSV/materialepas) works unchanged.

## Phases

- **Phase 0 — Spike (manual, no new code)**: hand-run the loop once. Export data with existing commands (`rux info --json`, `rux get components`, `rux export ply` + decimation, screenshots from `rux view`), drive Blender manually via blender-mcp with Claude, build one room's simplified model with tagged components. **Deliverable: a written assessment — does agent judgment actually beat the MIP pipeline on a real scan?** Kill or continue based on this.
- **Phase 1 — Agent eyes**: `rux render` subcommand — headless multi-view rendering (offscreen VTK/EGL, or a minimal custom rasterizer if VTK offscreen fights Nix) of clouds/meshes/labels to PNG: top-down orthographic ("floor plan"), 4 elevations, N perspective views; camera params CLI-controllable. Also useful independently (CI artifacts, docs, GUI thumbnails).
- **Phase 2 — rux MCP gateway**: Python FastMCP server (new `tools/rux-mcp/` or `bindings/mcp/`), wrapping `rux` subprocess calls; resources + tools per the diagram; Nix-packaged. Enable/extend pybind11 bindings where subprocess overhead hurts (bulk queries, full component geometry).
- **Phase 3 — Blender bridge + workflow**: evaluate community blender-mcp vs. official connector vs. blender-ai-mcp fork; GLB hand-off of detected plane patches / reference geometry; custom-property tagging convention (`reusex_guid`, `reusex_type`, `reusex_material_guid`); scripted scene-setup template (units, coordinate system matching ReUseX).
- **Phase 4 — Closed loop + write-back**: `save_components` write-back tool (schema: agent-generated components with provenance flag); round-trip validation metric; document the full agent playbook (prompt template + evidence package) so any agent session can run it.
- **Phase 5 — Evaluation**: same scans through (a) existing geometric pipeline, (b) agent workflow; compare component inventory precision/recall vs. manual ground truth, geometric accuracy, wall-clock and token cost. Decide the roles: replacement, fallback, or QA layer.

## Open decisions
- Gateway placement & packaging: `tools/rux-mcp/` Python package vs. under `bindings/`; stdio vs. HTTP transport (stdio first — local single-user, matches Claude Desktop/Code).
- Rendering backend: PCL/VTK offscreen (dep already present; EGL headaches on Nix possible) vs. minimal custom point rasterizer (more work, fully controlled) — spike in Phase 1.
- Blender connector choice (community/official/fork) — Phase 3 evaluation.
- Whether Astra specifically is worth testing as the modeling agent (API access, $10/$50 per M tokens) vs. Claude via Claude Code + MCP — Phase 5 comparison candidate.
- Relationship to ruxd: none for now (stays headless per #265); if the gateway ever needs remote operation, it becomes another client of the #265 API contract.

## Relation to other workstreams
- **#265 (GUI)**: complementary — GUI is the human's interface, this is the agent's. Both consume ProjectDB; both benefit from `rux render` (Phase 1 here reusable for GUI thumbnails). Human review/editing of agent output belongs in the GUI, not Blender.
- **#266 (DIRECTION.md)**: this is a new workstream ("agent-assisted modeling") — add it to the workstream table when #266 lands.
- Existing Blender addon (`apps/blender/`): superseded for this purpose; its pybind11 bindings are the reusable part. Do not extend the addon's per-frame loading approach.

## Non-goals
- Replacing the geometric pipeline outright before Phase 5 evidence exists.
- Human editing UX in Blender (GUI workstream).
- Remote/multi-user MCP deployment.
- Training/fine-tuning any model.

## Flagged uncertainties
- Astra blog specifics partially verified; treat its claims as directional, not spec.
- Official Blender MCP connector min version (4.2 vs 5.1) — verify in Phase 3.
- VTK offscreen rendering under Nix — verify early in Phase 1 (known EGL/OSMesa pitfalls).

**Comment by pfmephisto** [comment]
## Phase 2 (rux MCP gateway) is implemented — PR #349

`tools/rux-mcp/`, a read-only MCP gateway that shells out to the `rux` CLI.
Not merged, not closing this issue — Phases 3–5 are untouched.

### Tool surface shipped

| Tool | Backed by | Notes |
|---|---|---|
| `project_info()` | `rux info --json` | cheapest orientation call |
| `list_clouds()` | `rux info --json` | type, point count, label definitions |
| `list_meshes()` | `rux info --json` | vertex/face counts |
| `label_definitions(cloud)` | `rux info --json` | `{id: class name}` |
| `list_frames(limit, offset)` | `rux get frames` | paginated node ids |
| `get_frame(node_id)` | `rux get frames.ID` | intrinsics + 4×4 world pose |
| `list_components()` | `rux export csv` | **stopgap**, see gap 1 below |
| `get_component(id)` | `rux export csv` | by GUID or `component_name` |
| `list_passports()` | `rux info --json` | passport GUIDs |
| `get_passport(guid)` | `rux get materials.GUID` | full property set |
| `query_db(path)` | `rux get <path>` | allowlisted, JSON only |
| `render_view(view, layers, …)` | `rux render` | **image content** + paths |
| `analyze_quality(...)` | `rux analyze quality` | flatness RMS / thickness p90 |
| `analyze_accuracy(gt, ...)` | `rux analyze accuracy` | the loop's objective signal |
| `pipeline_log(limit)` | `rux log --json` | stage history |
| `validate_project(stage)` | `rux validate --json` | integrity or stage contract |

Resources: `rux://project/{summary,validation,components,log}`,
`rux://guide/query-paths`.

Mutating commands are out for v1. Write-back stays Phase 4.

### The progressive-disclosure rules are code, not documentation

`query_db` is deny-by-default over `rux get`. The binary paths — `clouds.NAME`
(raw PCD), `meshes.NAME.{data,texture,material}`,
`frames.ID.{color,depth,confidence}`, `labels.ID` — are refused *before* `rux`
is spawned, and the refusal names the way forward:

```
>>> query_db('clouds.cloud')
Error executing tool query_db: 'clouds.cloud' is not exposed: 'clouds.<name>'
streams the raw binary PCD. Use 'clouds.<name>.metadata' for the summary, or
render_view() to see the geometry.
```

The project path is bound at server startup, so no tool schema carries a path
an agent could redirect.

### Render example — the agent's eyes

```
>>> render_view(view='top', layers='cloud', size='800x600')
{"view": "top", "layers": ["cloud"], "size": "800x600",
 "image_count": 1, "paths": ["/tmp/rux-mcp-renders-…/view.png"]}
[image] image/png, 499564 base64 chars
```

Returned as MCP image content, not as coordinates. Committed as
`tools/rux-mcp/docs/demo-top-view.png` — a floor-plan view of the office scan
(`afb3234950`) that the gateway produced end to end via `rux render`.
`orbit:N` returns N numbered files, the first 4 inlined by default;
`inline=False` returns paths only.

### Gaps found while building it

These update the asset/gap table in the issue body:

1. **Gap 3 confirmed and narrower than stated.** Building components are not a
   `rux get` collection at all — the router registry
   (`apps/rux/src/database/resource_router.cpp`) implements `clouds, frames,
   labels, log, materials, meshes, panoramas, projects`. The `components` and
   `passports` entries in `rux get --help` are **stale**: passports live under
   `materials`, and components have no route. The gateway reads
   `rux export csv` (which opens the DB read-only) and parses component rows.
   Isolated in `rux_mcp/components.py` so it can be swapped for a real
   `rux get components` call. Component *geometry* is still unexposed;
   `render_view(layers='components')` is the only way to see where one sits.
   → worth its own issue: add a `components` router to `rux get`, and fix the
   stale help text.

2. **`rux` logs to stdout, not stderr.** A project one schema version behind
   the binary prints
   `[…] [rux] [warning] Project schema is v11 but this build expects v12`
   *ahead of* the JSON from `--json`, while stderr stays empty. Any consumer
   that pipes `rux … --json` into a parser breaks on this — it broke the
   gateway's integration test until `run_json()` learned to strip
   spdlog-formatted lines. The CLI applications install spdlog themselves
   (`apps/rux/src/rux.cpp`), so this is a one-line sink change.
   → also worth its own issue.

3. **MCP SDK friction worth knowing for Phase 3.** The Python SDK deliberately
   withholds the text of *unexpected* exceptions from the model: a tool raising
   `RuntimeError` reaches the agent as `Error executing tool X` and nothing
   more. Only `ToolError` (and argument-validation failures) carry their
   message through. Every tool here is wrapped accordingly — without it, none
   of the self-correcting refusal messages above would reach the agent. Also,
   a tool returning a bare `list` is exploded by the SDK into one content block
   per element, which is unreadable for a 16-element pose matrix; list-returning
   tools wrap in `{"count": …, "items": …}`.

### SDK decision

Official MCP Python SDK (`mcp>=2,<3`), not the standalone `fastmcp` package:
same decorator ergonomics (`fastmcp` v1 *was* `mcp.server.fastmcp`), one
dependency instead of two, tracks the spec directly. The 2.x rename
`FastMCP` → `MCPServer` is why the pin has an upper bound.

### Verification

90 unit tests green — `rux` replaced by a script that records argv and replays
canned stdout/stderr/exit status, tools driven through a real MCP client, with
the missing-project / non-zero-exit / timeout paths all covered. One
integration test passes against the office scan and skips cleanly when no
binary or project is available. `reuse lint` and pre-commit pass.

### Next

Phase 3 (Blender bridge) is unblocked from this side: the gateway can already
hand an agent the plane inventory, room labels, component list, quality metrics
and multi-view renders that the "agent evidence package" in the issue calls
for. The two `rux`-side gaps above are worth closing first so the component
half of that package stops going through a CSV.
